src/runtime/hexagon_dma.cpp

```diff
@@ -121,6 +127,150 @@ static void desc_pool_free (void* user_context) {
+        if ((dev->fmt == eDmaFmt_NV12)   ||
+            (dev->fmt == eDmaFmt_P010)   ||
+            (dev->fmt == eDmaFmt_NV124R) ||
+            (dev->fmt == eDmaFmt_TP10)) {
+            error(user_context) << "Hexagon: DMA pixel format not match Halide Buffer dimension \n";
+            return halide_error_code_device_buffer_copy_failed;
+        } 
+    }        
+ 
+    t_StDmaWrapper_RoiAlignInfo stWalkSize = {static_cast<uint16>(dst->dim[0].extent), static_cast<uint16>(dst->dim[1].extent)};
+    int nRet = nDmaWrapper_GetRecommendedWalkSize(dev->fmt, dev->is_ubwc, &stWalkSize);
+
+    roi_stride = nDmaWrapper_GetRecommendedIntermBufStride(currentFmt, &stWalkSize, dev->is_ubwc);
+    roi_width = stWalkSize.u16W;
+    roi_height = stWalkSize.u16H;
+
+    debug(user_context) << " roi_width " << roi_width;
+    debug(user_context) << " roi_height " << roi_height;
+    debug(user_context) << " roi_stride " << roi_stride;
+
+    // DMA driver Expect the Stride to be 256 Byte Aligned
+    halide_assert(user_context, (roi_stride % 256) == 0);
+
+    // Return NULL if descriptor is not allocated
+    void* desc_addr = desc_pool_get(user_context);
+    if (desc_addr == NULL) {
+        error(user_context) << "Hexagon: DMA descriptor allocation error \n";
+        return halide_error_code_device_buffer_copy_failed;
+    }
+
+    // Copy from Locked Cache to a temp DDR buffer
+    // TODO: This should be removed once the cache locking is addressed inside Halide Pipeline
+    int buf_size = roi_stride * roi_height * dst->type.bytes(); //dst->size_in_bytes();
+    debug(user_context) << " cache buffer size " << buf_size << "\n";
+    if (dev->cache_buf == 0) {
+        dev->cache_buf = HAP_cache_lock(buf_size, 0);
+    }
+
+    t_StDmaWrapper_DmaTransferSetup stDmaTransferParm;
+    stDmaTransferParm.eFmt                  = currentFmt; 
+    stDmaTransferParm.u16FrameW             = dev->frame_width;
+    stDmaTransferParm.u16FrameH             = dev->frame_height;
+    stDmaTransferParm.u16FrameStride        = dev->frame_stride;
+    stDmaTransferParm.u16RoiW               = roi_width;
+    stDmaTransferParm.u16RoiH               = roi_height;
+    stDmaTransferParm.u16RoiStride          = roi_stride;
+    stDmaTransferParm.bIsFmtUbwc            = dev->is_ubwc;
+    stDmaTransferParm.bUse16BitPaddingInL2  = 0;
+    stDmaTransferParm.pDescBuf              = desc_addr;
+    stDmaTransferParm.pTcmDataBuf           = dev->cache_buf;
+    stDmaTransferParm.pFrameBuf             = dev->buffer;
+    if (dev->is_write) {
+       stDmaTransferParm.eTransferType      = eDmaWrapper_L2ToDdr;
+    } else {
+       stDmaTransferParm.eTransferType      = eDmaWrapper_DdrToL2;
+    }
+    stDmaTransferParm.u16RoiX               = dev->offset_x + dst->dim[0].min;
+    stDmaTransferParm.u16RoiY               = dev->offset_y + dst->dim[1].min;
+
+    // DMA Driver Halves the Height and Y Offset so that only Half the ROI Size of Luma is transferred  for chroma
```

Comment:
Something I'm still confused about: does the DMA hardware do anything with the information that these planes are (Y)UV data? i.e. does it do chroma upsampling/downsampling, or RGB <-> YUV conversion?

Reply:
DMA hardware does the down sampling automatically when we choose the format as eDmaFmt_NV12_UV

Reply:
OK, so one would only use the YUV formats when one actually wants the DMA hardware to do a conversion, right? If just processing YUV data without conversion, we would just use the basic data types?